prov/rstream/src/rstream_msg.c

```diff
@@ -0,0 +1,571 @@
+    assert(len_available <= mr->avail_size);
+    if (!len_available)
+        return 0;
+
+    *data_addr = *data_addr + mr->start_offset;
+    uint32_t len = (len_available <  req_len) ? len_available : req_len;
+    assert(mr->avail_size >= len);
+    mr->avail_size = mr->avail_size - len;
+    mr->start_offset = (mr->start_offset + len) % mr->size;
+    return len;
+}
+
+static void rstream_return_contig_len(struct rstream_mr_seg *mr, uint32_t len)
+{
+    assert((mr->avail_size + len) <= mr->size);
+    mr->avail_size = mr->avail_size + len;
+    mr->end_offset = (mr->end_offset + len) % mr->size;
+}
+
+static int rstream_check_for_rx_comp(struct rstream_ep *ep,
+    struct fi_cq_data_entry *completion_entry);
+
+static int rstream_send_ctrl_msg(struct rstream_ep *ep, uint64_t imm_data)
+{
+    if (!ep->qp_win.ctrl_credits || !ep->qp_win.target_rx_credits) {
+        if (!rstream_check_for_tx_comp(ep) || rstream_target_rx_full(ep)){
+            FI_DBG(&rstream_prov, FI_LOG_EP_CTRL,
+                "ctrl msg exhaustion \n");
+           return 0;
+        }
+    }
+
+    if (RSTREAM_USING_IWARP) {
+        int ret = fi_inject(ep->ep_fd, &imm_data, RSTREAM_IWARP_DATA_SIZE, 0);
+        if (ret != 0)
+            return 0;
+    } else {
+        const struct fi_msg msg =
+            {.msg_iov = NULL, .desc = NULL, .iov_count = 0,
+                .context = rstream_get_tx_ctx(ep), .data = imm_data};
+        int ret = fi_sendmsg(ep->ep_fd, &msg, FI_REMOTE_CQ_DATA);
+        if (ret != 0)
+            return 0;
+
+        if (ep->qp_win.tx_credits > 0) {
+            ep->qp_win.tx_credits = ep->qp_win.tx_credits - 1;
+        } else {
+            ep->qp_win.ctrl_credits = ep->qp_win.ctrl_credits - 1;
+            rstream_check_for_tx_comp(ep);
+        }
+    }
+    assert(ep->qp_win.target_rx_credits > 0);
+    ep->qp_win.target_rx_credits = ep->qp_win.target_rx_credits - 1;
+
+    return 1;
+}
+
+/* accumulate data in tx_cq exhaustion case */
+static void rstream_update_target(struct rstream_ep *ep, uint16_t num_completions,
+    uint32_t len)
+{
+    uint64_t imm_data = 0;
+
+    ep->imm_data.num_completions =
+        ep->imm_data.num_completions + num_completions;
+    ep->imm_data.total_len = ep->imm_data.total_len + len;
+
+    if (ep->imm_data.num_completions)
+        rstream_imm_set_credits(&imm_data, ep->imm_data.num_completions);
+    if (ep->imm_data.total_len)
+        rstream_imm_set_len(&imm_data, ep->imm_data.total_len);
+
+    if ((ep->imm_data.num_completions >= ep->qp_win.max_rx_credits / 2) ||
+        (ep->imm_data.total_len >= ep->local_mr.rx.size / 2)) {
+        if (rstream_send_ctrl_msg(ep, imm_data)) {
+            FI_DBG(&rstream_prov, FI_LOG_EP_CTRL,
+                "ctrl msg update %u = completions %u = len \n",
+                ep->imm_data.num_completions, ep->imm_data.total_len);
+            ep->imm_data.num_completions = 0;
+            ep->imm_data.total_len = 0;
```

Comment:
The common function behavior is to return 0 on success and non-zero (-errno) on error.  Does rstream_send_ctrl_msg() (and other rstream calls) follow this model?